### ISSUE TITLE
Game::drawTextScreen(): StringIndexOutOfBoundsException 

### ISSUE THREAD

**Issue by dgant** [JavaBWAPI/JBWAPI#54]
Stack trace:
```
stackTrace = {StackTraceElement[24]@6004} 
 0 = {StackTraceElement@6005} "bwapi.WrappedBuffer.putString(WrappedBuffer.java:80)"
 1 = {StackTraceElement@6006} "bwapi.ClientData$GameData.setStrings(ClientData.java:562)"
 2 = {StackTraceElement@6007} "bwapi.Client.addString(Client.java:251)"
 3 = {StackTraceElement@6008} "bwapi.Game.drawText(Game.java:1738)"
 4 = {StackTraceElement@6009} "bwapi.Game.drawTextScreen(Game.java:1760)"
 5 = (bot code)
```

The input string was of length 2920:
```
L: PvP Jadien @ (4)Gladiator1.1.scx (PvPRobo, Fingerprint2Gate, FingerprintDragoonRange, FingerprintGatewayFirst, FingerprintMannerPylon, FingerprintProxyGateway)
W: PvP Jadien @ (3)Power Bond.scx (PvP2GateDTExpand)
W: PvP Jadien @ (2)Overwatch(n).scx (PvP2GateDTExpand, Fingerprint1GateCore, Fingerprint4GateGoon, FingerprintDragoonRange, FingerprintGatewayFirst, FingerprintMannerPylon)
L: PvP Jadien @ (4)Fighting Spirit.scx (PvPProxy2Gate, Fingerprint1GateCore, Fingerprint2Gate, FingerprintDragoonRange, FingerprintGatewayFirst, FingerprintMannerPylon)
L: TvT Atlas Wing @ (2)Overwatch(n).scx (TvT1FacFE, TvT2BaseBC, FingerprintBio)
W: TvT Epsilon Squadron @ (4)Gladiator1.1.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Epsilon Squadron @ (4)Fighting Spirit.scx (TvT2FacTanks, TvT5Fac, Fingerprint2Rax1113, FingerprintBio)
W: TvT Mar Sara @ (2)Tres Pass.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Epsilon Squadron @ (4)Fighting Spirit.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Antiga @ (4)CircuitBreakers1.0.scx (TvT1RaxFE, TvT5Fac, FingerprintBio)
W: TvT Epsilon Squadron @ (4)CircuitBreakers1.0.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Elite Guard @ (2)Overwatch(n).scx (TvT1FacFE, TvT2BaseBC, FingerprintBBS, FingerprintBio)
W: TvT Delta Squadron @ (4)Fighting Spirit.scx (TvT1RaxFE, TvT2Base2Port, TvT2BaseBC, FingerprintBBS, FingerprintBio)
W: TvT Mar Sara @ (2)Tres Pass.scx (TvT1FacPort, TvT5Fac, FingerprintBBS, FingerprintBio)
W: TvT Delta Squadron @ (3)Power Bond.scx (TvT1FacFE, TvT2Base2Port, TvT2BaseBC, FingerprintBio)
L: TvT Delta Squadron @ (2)Overwatch(n).scx (TvE1RaxSCVMarine, FingerprintBio)
W: TvT Epsilon Squadron @ (4)Fighting Spirit.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Cronus Wing @ (2)Overwatch(n).scx (TvT2Port, TvT2Base2Port, TvT2BaseBC, FingerprintBio)
W: TvT Epsilon Squadron @ (4)CircuitBreakers1.0.scx (TvT2FacTanks, TvT5Fac, FingerprintBio)
W: TvT Antiga @ (3)Power Bond.scx (TvT1RaxFE, TvT5Fac, FingerprintBio)
W: TvT Cronus Wing @ (4)Gladiator1.1.scx (TvT2FacTanks, TvT2Base2Port, TvT2BaseBC, FingerprintBBS, FingerprintBio)
L: TvT Kel-Morian Combine @ (4)Gladiator1.1.scx (TvT14CC, TvT5Fac)
W: TvT Cronus Wing @ (4)Fighting Spirit.scx (TvT1FacPort, TvT2Base2Port, TvT2BaseBC, FingerprintBBS, FingerprintBio)
W: TvT Cronus Wing @ (2)Overwatch(n).scx (TvT2FacTanks, TvT2Base2Port, TvT2BaseBC, FingerprintBio)
W: TvT Kel-Morian Combine @ (4)CircuitBreakers1.0.scx (TvE2RaxSCVMarine, FingerprintBio)
W: TvT Atlas Wing @ (4)Fighting Spirit.scx (TvT1FacFE, TvT2BaseBC, FingerprintBBS, FingerprintBio)
W: TvT Epsilon Squadron @ (2)Tres Pass.scx (TvT2FacTanks, TvT5Fac, FingerprintBio, FingerprintWorkerRush)
W: TvT Antiga @ (4)Gladiator1.1.scx (TvT1RaxFE, TvT5Fac, Fingerprint2Rax1113, FingerprintBio)
L: TvT Delta Squadron @ (4)CircuitBreakers1.0.scx (TvEProxyBBS)
L: TvT Cronus Wing @ (4)Fighting Spirit.scx (TvEProxyBBS, FingerprintBio)
```

**Comment by dgant** [comment]
So the cause is pretty straightforward:
```
void putString(final int offset, final int maxLen, final String string) {
        if (string.length() + 1 >= maxLen) {
            throw new StringIndexOutOfBoundsException();
        }
```
I propose, mostly in ignorance of the considerations involved, that a more elegant handling would be to just truncate any strings to the limit.

**Comment by dgant** [comment]
Proposed alternative implementation:
```
void putString(final int offset, final int maxLen, final String string) {
        String finalString = string.length() +1 >= maxLen ? string.substring(0, maxLen) : string;
        long pos = offset + address;
        for (int i = 0; i < finalString.length(); i++) {
            unsafe.putByte(pos, (byte) finalString.charAt(i));
            pos++;
        }
        unsafe.putByte(pos, (byte) 0);
    }
```

**Comment by N00byEdge** [comment]
Or the string could be split up into multiple individually drawn parts

**Comment by JasperGeurtz** [comment]
![2020-02-21 22_49_54](https://user-images.githubusercontent.com/10107858/75074760-a716c300-54fc-11ea-87f0-5989cdde6ce8.png)

BWAPI & BWMirror indeed seem to cut it off

**Comment by JasperGeurtz** [comment]
Fixed in a9e55ec
